Button-less-login

klyngen opened the pull request:
Pressing login button is no longer required
Also transitioning over to axios with a neat httpClient that has a
middleware resolving the bearer-token

Commit:
Better login
Pressing login button is no longer required
Also transitioning over to axios with a neat httpClient that has a
middleware resolving the bearer-token

palbock commented:
Ser ikke navnet til innlogget bruker i hamburgermenyen:
![image](https://user-images.githubusercontent.com/41721443/132003781-171b37fb-c2a6-4a9d-a0be-c90442f6135c.png)

klyngen commented:
> Ser ikke navnet til innlogget bruker i hamburgermenyen:
> ![image](https://user-images.githubusercontent.com/41721443/132003781-171b37fb-c2a6-4a9d-a0be-c90442f6135c.png)

Det er fordi du er logget inn som Ahre Ketil. Om du er logget inn med din faktiske bruker, så vil navnet vises :) med andre ord. Om du logger inn med azure så får du et navn :)

Trulsaa commented on packages/frontend/src/services/auth.ts:
```diff
@@ -1,106 +1,129 @@
-// https://docs.microsoft.com/en-us/azure/active-directory/develop/msal-js-initializing-client-applications
+const scopes = [config.ACCESS_SCOPE];
```
Kunne vi ha lagt denne inne i klassen? Den ser så malplasert ut.

Trulsaa commented on packages/frontend/src/services/auth.ts:
```diff
@@ -1,106 +1,129 @@
-  const authHeaders = { Authorization: `Bearer ${accessToken}` };
+  public isLoggedIn(): boolean {
+    return this.msalManager.getAllAccounts()[0] == null;
+  }
```
Denne hadde veldig rar logikk :P

Trulsaa commented on packages/frontend/src/services/auth.ts:
```diff
@@ -1,106 +1,129 @@
+  EventMessage,
+} from "@azure/msal-browser";
+export class AuthService {
+  private accountLoaded$: Promise<AccountInfo>;
```
Jeg syntes AccountInfor bør være lagret i kun 1 variabel. Altså omgjøre acountsLoaded til Promise<boolean> og alltid lagre AccountInfo i accountInfo. Eller noe lignende det. accountLoaded blir i allefall et misvisende navn om den ikke returnerer en boolean.

Trulsaa commented on packages/frontend/src/store/tasks.ts:
```diff
@@ -58,47 +58,34 @@ const mutations = {
-      const url = new URL(config.API_HOST + "/api/user/tasks").toString();
-      const res = await adAuthenticatedFetch(url);
-      if (res.status === 404) {
-        commit("SET_USER_NOT_FOUND");
```
Blir digg å få flyttet denne inn i en middleware :) Ser den har blitt kopiert rundt i hele koden nå 😂

Trulsaa commented on packages/frontend/src/store/timeEntries.ts:
```diff
@@ -79,30 +79,26 @@ const actions = {
+            );
+          })
+          .catch(e => {
+            if (e.status !== 200) {
```
Denne if statementen kan vel aldri bli true.

Trulsaa commented on packages/frontend/src/store/timeEntries.ts:
```diff
@@ -79,30 +79,26 @@ const actions = {
+            }
+            if (!Array.isArray(e.data) && e.data.message) {
+              throw Error(e.data.message);
+            }
```
Jeg vil tro dette er noe som henger igjen fra gamle dager da APIet ikke var helt 👍 Kanskje gjøre en liten test og sjekke om det fortsatt trengs?

Trulsaa commented on packages/frontend/src/store/timeEntries.ts:
```diff
@@ -79,30 +79,26 @@ const actions = {
 
       commit("FLUSH_PUSH_QUEUE");
       try {
```
Trenger vi fortsatt denne try catch blocka med all disse flotterse then().catch() lenkene 😎

Trulsaa commented on packages/frontend/src/store/timeEntries.ts:
```diff
@@ -118,20 +114,21 @@ ${timeEntries.title}`);
+    await httpClient
+      .get(url.toString())
+      .then(response => {
+        if (!Array.isArray(response.data) && response.data.message) {
```
Trenger vi virkelig denne lenger? Jeg har lyst til å tro att dette er noe som henger igjen fra gamle dager.

Trulsaa commented:
<img width="874" alt="Screenshot 2021-09-03 at 20 54 01" src="https://user-images.githubusercontent.com/15033235/132053678-50161180-ee78-41a9-a9c5-87fbdce91fa9.png">
Jeg kommer ikke inn i appen og får ☝️ i consolen når jeg prøver å teste dette etter å ha satt isDevMode til false.

FredrikAugust commented on packages/frontend/src/store/timeEntries.ts:
```diff
@@ -79,30 +79,26 @@ const actions = {
-          );
-        }
+        await httpClient
+          .post<any[]>(
```
Kan denne types til en type? 

Hvis det er vanskelig/vil endres veldig ofte og du stoler på at "createTimeEntrie"-inputet matcher det som kommer fra backend går det også an å type det til input-et til `createTimeEntrie`;

```ts
Array<Parameters<typeof createTimeEntrie>[0]>
```

Men det er kanskje ikke det klareste eller mest leselige i verden

FredrikAugust replied:
Jeg ser nå at parameterene til `createTimeEntrie` er `any` også. Kanskje vi kan type det vi vet vi trenger ved å endre på `createTimeEntrie`? 

```ts
function createTimeEntrie<T extends { date: string, value: number }>(data: T): FrontendTimentrie {
```

og 

```suggestion
          .post<Array<Parameters<typeof createTimeEntrie>[0]>>(
```

Det vil i hvert fall være en start på å få solid typing

FredrikAugust commented on packages/frontend/src/store/tasks.ts:
```diff
@@ -58,47 +58,34 @@ const mutations = {
+      });
   },
 
   PUSH_TASKS: async (
```
Denne trenger ikke `async`-keywordet hvis den ikke skal bruke `async/await`, det holder at den bruker `Promise.then`. Evt. kan den skrives om til å bruke `try/catch` rundt `async/await`-kall.

```suggestion
  PUSH_TASKS: (
```

FredrikAugust commented on packages/frontend/src/store/overtime.ts:
```diff
@@ -264,39 +242,31 @@ const actions = {
-          date: parameters.date,
-          hours: parameters.hours,
-        }),
+    await httpClient
```
Trengs denne å awaites (og funksjonen å være `async`) når vi bruker `Promise.then`?

Commit:
Better errorhandling
Removed error-handling from most store code
Most fetch-request had previously a catch-block waiting for a
404-status. All that code is now in a middleware and error-messages are
much easier to use.

Commit:
Removed more legacy error handling-code

Commit:
Merge
Merge branch 'master' of https://github.com/Alv-no/alvtime into feature/no_login_button

klyngen committed:
Update packages/frontend/src/store/timeEntries.ts

Ehm. Ja  kunne heeelt sikkert vært typer, men det er ikke det denne PR'en handler om. Jeg har bare ryddet opp i request-koden. 

Det burde være mere typer i frontend, men det kan introduseres i en annen PR

Co-authored-by: Fredrik August Madsen-Malmo <fredrik.malmo@icloud.com>

Commit:
Removed out of place async/await code

Commit:
Removed token-list bugs

Commit:
Added back devmode

Trulsaa commented on packages/frontend/src/components/GenerateToken.vue:
```diff
@@ -78,28 +78,19 @@ export default Vue.extend({
     },
 
     async fetchAccessToken() {
```
async trengs vel ikke lenger

Trulsaa commented on packages/frontend/src/components/FavoriteSelector.vue:
```diff
@@ -64,6 +64,9 @@ export default Vue.extend({
     },
   },
+  async created() {
+    await this.$store.dispatch("FETCH_TASKS");
```
async / await trengs vel ikke her

Commit:
Removed read code

Trulsaa commented on packages/frontend/src/components/TimeEntrieDayList.vue:
```diff
@@ -88,7 +87,9 @@ export default Vue.extend({
     },
   },
-
+  async created() {
```
her trengs vel ikke async

Trulsaa commented on packages/frontend/src/components/TimeEntrieWeekList.vue:
```diff
@@ -95,6 +95,9 @@ export default Vue.extend({
     },
   },
 
+  async created() {
```
her trengs vel ikke async

Commit:
Resolved PR-comments

Trulsaa commented on packages/frontend/src/components/ErrorSnackbar.vue:
```diff
@@ -62,8 +74,14 @@ export default Vue.extend({
+    },
+    getLastIssue(): ErrorResponse {
+      return this.issues[this.issues.length - 1];
+    },
```
Disse ser ikke ut til å brukes noe sted

Trulsaa commented on packages/frontend/src/components/TimeEntrieDayList.vue:
```diff
@@ -88,7 +87,9 @@ export default Vue.extend({
-
+  created() {
+    this.$store.dispatch("FETCH_TASKS");
+  },
```
Kan denne flyttes opp i Hours.vue?

Trulsaa commented on packages/frontend/src/components/TimeEntrieWeekList.vue:
```diff
@@ -95,6 +95,9 @@ export default Vue.extend({
     },
   },
 
+  created() {
+    this.$store.dispatch("FETCH_TASKS");
+  },
```
Kan denne flyttes opp i Hours.vue?

Trulsaa commented on packages/frontend/src/services/httpClient.ts:
```diff
@@ -0,0 +1,91 @@
+    };
+
+    // We should never do a request without a token therefore a 401
+    // means that we used a token and the token was not accepted
```
På linje 42 settes det på et token på de urlene som er protected og ellers ikke. Altså blir vel denne kommentaren feil?

Trulsaa reviewed (dismissed):
Her ble det mange endringer som må testes grundig før det merges. Jeg skulle helst sett at dette ble splittet opp i to PRer. En for Login og en for Error håndtering. På den måten kan man få gjennomført kodereview og testing mye raskere.

Commit:
More general error-handling. Tested error messages from various 404, 401 and 500 errors. Tested through provoking errors for trying to flex more hours than possible and by extracting too many hours for payment

Commit:
Resolved last comments

Commit:
Removed garbage from code

FredrikAugust reviewed (commented):
lgtm
